File: golang/run-tests.go
// Power Laws & Fractals - Go Test Runner
//
// Follows the unified testing protocol:
// 1. Load base-data.json (for systems configuration + base scales)
// 2. Load test-input.json (raw facts only)
// 3. Compute derived values for test scales
// 4. Output results to test-results/golang-results.json
// 5. Validate against answer-key.json
// 6. Display with unified visualization (all 8 iterations, colors, ASCII plots)

package main

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"erb-power-laws/pkg/rulebook"
)

// ANSI colors
const (
	green   = "\033[92m"
	yellow  = "\033[93m"
	cyan    = "\033[96m"
	red     = "\033[91m"
	dim     = "\033[2m"
	reset   = "\033[0m"
	bold    = "\033[1m"
	magenta = "\033[95m"
)

// Plot characters
const (
	plotActual      = "●"
	plotProjected   = "◌"
	plotTheoretical = "·"
)

func main() {
	// Find project root (parent of golang directory)
	execPath, _ := os.Getwd()
	projectRoot := filepath.Dir(execPath)
	
	// If we're already at project root (golang is a subdirectory)
	if _, err := os.Stat(filepath.Join(execPath, "test-data")); err == nil {
		projectRoot = execPath
	}
	
	// Paths
	testDataDir := filepath.Join(projectRoot, "test-data")
	testResultsDir := filepath.Join(projectRoot, "test-results")
	
	baseDataPath := filepath.Join(testDataDir, "base-data.json")
	testInputPath := filepath.Join(testDataDir, "test-input.json")
	answerKeyPath := filepath.Join(testDataDir, "answer-key.json")
	resultsPath := filepath.Join(testResultsDir, "golang-results.json")
	
	// Ensure results directory exists
	os.MkdirAll(testResultsDir, 0755)
	
	// Load base data
	baseData, err := rulebook.LoadBaseData(baseDataPath)
	if err != nil {
		fmt.Printf("%sError: Could not load base-data.json: %v%s\n", red, err, reset)
		os.Exit(1)
	}
	
	// Load test input
	testInput, err := rulebook.LoadTestInput(testInputPath)
	if err != nil {
		fmt.Printf("%sError: Could not load test-input.json: %v%s\n", red, err, reset)
		os.Exit(1)
	}
	
	// Load answer key
	answerKey, err := rulebook.LoadAnswerKey(answerKeyPath)
	if err != nil {
		fmt.Printf("%sError: Could not load answer-key.json: %v%s\n", red, err, reset)
		os.Exit(1)
	}
	
	// Build systems map
	systemsMap := rulebook.BuildSystemsMap(baseData.Systems)
	
	// Compute derived values for test scales
	computedTestScales := make([]map[string]interface{}, 0, len(testInput.Scales))
	
	for i := range testInput.Scales {
		scale := &testInput.Scales[i]
		scale.CalculateAllFields(systemsMap)
		computedTestScales = append(computedTestScales, scale.ToOutputMap())
	}
	
	// Save results (test scales only for validation)
	results := &rulebook.TestResults{
		Platform:  "golang",
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Scales:    computedTestScales,
	}
	
	err = rulebook.SaveResults(resultsPath, results)
	if err != nil {
		fmt.Printf("%sError: Could not save results: %v%s\n", red, err, reset)
		os.Exit(1)
	}
	
	// Merge base scales with computed test scales for full visualization
	allScales := mergeScales(baseData.Scales, computedTestScales, systemsMap)
	
	// Validate against answer key
	passCount, failCount, failures := rulebook.ValidateAllScales(computedTestScales, answerKey)
	
	// Print full report
	printFullReport(systemsMap, allScales, passCount, failCount, failures)
	
	// Exit with appropriate code
	if failCount > 0 {
		os.Exit(1)
	}
}

// mergeScales combines base scales with computed test scales
func mergeScales(baseScales []rulebook.Scale, testScales []map[string]interface{}, systems rulebook.SystemsMap) []map[string]interface{} {
	// Convert base scales to maps
	all := make([]map[string]interface{}, 0, len(baseScales)+len(testScales))
	
	for i := range baseScales {
		scale := &baseScales[i]
		scale.CalculateAllFields(systems)
		all = append(all, scale.ToOutputMap())
	}
	
	// Add test scales
	all = append(all, testScales...)
	
	return all
}

// renderASCIIPlot creates an ASCII log-log plot
func renderASCIIPlot(scales []map[string]interface{}, system *rulebook.System, width, height int) string {
	if len(scales) == 0 {
		return "  (No data)"
	}
	
	// Extract points
	type point struct {
		x, y        float64
		isProjected bool
	}
	
	var points []point
	for _, s := range scales {
		logScale, ok1 := s["LogScale"].(float64)
		logMeasure, ok2 := s["LogMeasure"].(float64)
		isProj, _ := s["IsProjected"].(bool)
		if ok1 && ok2 {
			points = append(points, point{logScale, logMeasure, isProj})
		}
	}
	
	if len(points) == 0 {
		return "  (No valid data points)"
	}
	
	// Calculate bounds
	xMin, xMax := points[0].x, points[0].x
	yMin, yMax := points[0].y, points[0].y
	for _, p := range points {
		if p.x < xMin { xMin = p.x }
		if p.x > xMax { xMax = p.x }
		if p.y < yMin { yMin = p.y }
		if p.y > yMax { yMax = p.y }
	}
	
	xRange := xMax - xMin
	if xRange == 0 { xRange = 1 }
	yRange := yMax - yMin
	if yRange == 0 { yRange = 1 }
	
	// Create grid
	grid := make([][]string, height)
	for i := range grid {
		grid[i] = make([]string, width)
		for j := range grid[i] {
			grid[i][j] = " "
		}
	}
	
	// Map to grid coordinates
	toGrid := func(x, y float64) (int, int) {
		gx := int((x - xMin) / xRange * float64(width-1))
		gy := height - 1 - int((y - yMin) / yRange * float64(height-1))
		if gx < 0 { gx = 0 }
		if gx >= width { gx = width - 1 }
		if gy < 0 { gy = 0 }
		if gy >= height { gy = height - 1 }
		return gx, gy
	}
	
	// Draw theoretical slope line
	slope := system.TheoreticalLogLogSlope
	if slope != 0 {
		x0, y0 := points[0].x, points[0].y
		for i := 0; i < width; i++ {
			x := xMin + (float64(i) / float64(width-1)) * xRange
			y := y0 + slope * (x - x0)
			if y >= yMin && y <= yMax {
				gx, gy := toGrid(x, y)
				if grid[gy][gx] == " " {
					grid[gy][gx] = dim + plotTheoretical + reset
				}
			}
		}
	}
	
	// Sort: actual first, then projected (so projected overlays)
	sort.Slice(points, func(i, j int) bool {
		return !points[i].isProjected && points[j].isProjected
	})
	
	// Plot points
	for _, p := range points {
		gx, gy := toGrid(p.x, p.y)
		if p.isProjected {
			grid[gy][gx] = magenta + plotProjected + reset
		} else {
			grid[gy][gx] = green + plotActual + reset
		}
	}
	
	// Build output
	var lines []string
	
	lines = append(lines, fmt.Sprintf("  %slog(Measure)%s", dim, reset))
	lines = append(lines, fmt.Sprintf("  %7.2f ┤", yMax))
	
	for i, row := range grid {
		prefix := "        │"
		if i == len(grid)-1 {
			prefix = fmt.Sprintf("  %7.2f ┤", yMin)
		}
		lines = append(lines, prefix + strings.Join(row, ""))
	}
	
	lines = append(lines, fmt.Sprintf("         └%s", strings.Repeat("─", width)))
	lines = append(lines, fmt.Sprintf("         %-7.2f%s%7.2f", xMin, strings.Repeat(" ", width-14), xMax))
	lines = append(lines, fmt.Sprintf("  %s%s%s", dim, center("log(Scale)", width+9), reset))
	lines = append(lines, fmt.Sprintf("  %s●%s Actual   %s◌%s Projected   %s·%s Theoretical (slope=%.3f)", 
		green, reset, magenta, reset, dim, reset, slope))
	
	return strings.Join(lines, "\n")
}

func center(s string, width int) string {
	if len(s) >= width {
		return s
	}
	padding := (width - len(s)) / 2
	return strings.Repeat(" ", padding) + s + strings.Repeat(" ", width-len(s)-padding)
}

func printSystemTable(scales []map[string]interface{}, system *rulebook.System) {
	icon := "📈"
	if system != nil && system.Class == "fractal" {
		icon = "🔺"
	}
	
	displayName := system.SystemID
	if system != nil {
		displayName = system.DisplayName
	}
	
	fmt.Printf("\n%s %s%s%s\n", icon, bold, displayName, reset)
	fmt.Printf("  %sTheoretical slope: %.3f%s\n", dim, system.TheoreticalLogLogSlope, reset)
	
	fmt.Printf("\n  %4s  %12s  %14s  %10s  %12s  %10s\n", "Iter", "Measure", "Scale", "LogScale", "LogMeasure", "Type")
	fmt.Println("  " + strings.Repeat("─", 70))
	
	// Sort by iteration
	sort.Slice(scales, func(i, j int) bool {
		return scales[i]["Iteration"].(int) < scales[j]["Iteration"].(int)
	})
	
	for _, s := range scales {
		isProj, _ := s["IsProjected"].(bool)
		color := green
		marker := "●"
		typeLabel := "actual"
		if isProj {
			color = magenta
			marker = "◌"
			typeLabel = "projected"
		}
		
		fmt.Printf("  %s%4d  %12.6f  %14.8f  %10.5f  %12.5f  %s %s%s\n",
			color,
			s["Iteration"].(int),
			s["Measure"].(float64),
			s["Scale"].(float64),
			s["LogScale"].(float64),
			s["LogMeasure"].(float64),
			marker,
			typeLabel,
			reset)
	}
	
	fmt.Printf("\n  %sRow count: %d%s\n", dim, len(scales), reset)
}

func printFullReport(systems rulebook.SystemsMap, allScales []map[string]interface{}, 
	passCount, failCount int, failures []rulebook.ValidationResult) {
	
	fmt.Printf("\n%s================================================================================\n", bold)
	fmt.Printf("  🐹 POWER LAWS & FRACTALS - Go Test Runner%s\n", reset)
	fmt.Printf("%s================================================================================\n", reset)
	
	fmt.Printf("\n%sAll Computed Values (from Go):%s\n", cyan, reset)
	fmt.Printf("  %s●%s Green = Actual Data (iterations 0-3)\n", green, reset)
	fmt.Printf("  %s◌%s Magenta = Projected/Computed (iterations 4-7)\n", magenta, reset)
	fmt.Println(strings.Repeat("─", 80))
	
	// Group scales by system
	bySystem := make(map[string][]map[string]interface{})
	for _, scale := range allScales {
		systemID := scale["System"].(string)
		bySystem[systemID] = append(bySystem[systemID], scale)
	}
	
	// Get sorted system IDs
	systemIDs := make([]string, 0, len(bySystem))
	for id := range bySystem {
		systemIDs = append(systemIDs, id)
	}
	sort.Strings(systemIDs)
	
	for _, systemID := range systemIDs {
		scales := bySystem[systemID]
		system := systems[systemID]
		
		// Print table
		printSystemTable(scales, system)
		
		// Print ASCII plot
		fmt.Printf("\n%s  Log-Log Plot:%s\n", cyan, reset)
		plot := renderASCIIPlot(scales, system, 50, 12)
		fmt.Println(plot)
	}
	
	// Validation results
	fmt.Printf("\n%s================================================================================\n", reset)
	fmt.Printf("%sValidation Results (projected scales vs answer-key):%s\n", cyan, reset)
	fmt.Println(strings.Repeat("─", 80))
	
	if failCount == 0 {
		fmt.Printf("  %s✓ All %d projected scales validated successfully!%s\n", green, passCount, reset)
	} else {
		fmt.Printf("  %s⚠ %d passed, %d failed%s\n", yellow, passCount, failCount, reset)
		for i, failure := range failures {
			if i >= 5 {
				break
			}
			fmt.Printf("    • %s:\n", failure.ScaleID)
			for _, m := range failure.Mismatches {
				fmt.Printf("      - %s\n", m)
			}
		}
	}
	
	// Summary
	totalScales := len(allScales)
	actualCount := 0
	projectedCount := 0
	for _, s := range allScales {
		if isProj, ok := s["IsProjected"].(bool); ok && isProj {
			projectedCount++
		} else {
			actualCount++
		}
	}
	
	fmt.Printf("\n%s================================================================================\n", reset)
	fmt.Printf("  %sSummary:%s\n", bold, reset)
	fmt.Printf("    Systems: %d\n", len(bySystem))
	fmt.Printf("    Total scales: %d (%d per system)\n", totalScales, totalScales/len(bySystem))
	fmt.Printf("    Actual (0-3): %d\n", actualCount)
	fmt.Printf("    Projected (4-7): %d\n", projectedCount)
	fmt.Println("================================================================================")
	fmt.Printf("  %s✓ Go test run complete!%s\n", green, reset)
	fmt.Println("================================================================================\n")
}

File: python/run-tests.py
#!/usr/bin/env python3
"""
Power Laws & Fractals - Python Test Runner

Follows the unified testing protocol:
1. Load base-data.json (for systems configuration + base scales)
2. Load test-input.json (raw facts only)
3. Compute derived values for test scales
4. Output results to test-results/python-results.json
5. Validate against answer-key.json
6. Display with unified visualization (all 8 iterations, colors, ASCII plots)
"""

import json
import math
import sys
from pathlib import Path
from datetime import datetime, timezone
from dataclasses import dataclass
from typing import Dict, List, Optional

# Add visualizer to path for shared library
SCRIPT_DIR = Path(__file__).parent
PROJECT_ROOT = SCRIPT_DIR.parent
sys.path.insert(0, str(PROJECT_ROOT / 'visualizer'))

from console_output import print_full_report, merge_scales

# Paths
TEST_DATA_DIR = PROJECT_ROOT / 'test-data'
TEST_RESULTS_DIR = PROJECT_ROOT / 'test-results'

# ANSI colors (for error messages before shared lib loads)
RED = '\033[91m'
RESET = '\033[0m'

# Tolerance for validation (allows for floating-point precision in 6dp comparisons)
# Using 0.0000015 to handle rounding at the 6th decimal place boundary
TOLERANCE = 0.0000015


@dataclass
class System:
    """System configuration from ERB"""
    SystemID: str
    DisplayName: str
    Class: str
    BaseScale: float
    ScaleFactor: float
    MeasureName: str
    FractalDimension: Optional[float]
    TheoreticalLogLogSlope: float


@dataclass
class Scale:
    """Scale measurement with computed values"""
    ScaleID: str
    System: str
    Iteration: int
    Measure: float
    BaseScale: Optional[float] = None
    ScaleFactor: Optional[float] = None
    ScaleFactorPower: Optional[float] = None
    Scale: Optional[float] = None
    LogScale: Optional[float] = None
    LogMeasure: Optional[float] = None
    IsProjected: bool = False


def load_json(path: Path) -> Dict:
    """Load JSON file"""
    with open(path, 'r') as f:
        return json.load(f)


def save_json(path: Path, data: Dict):
    """Save JSON file"""
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, 'w') as f:
        json.dump(data, f, indent=2)


def load_systems(base_data: Dict) -> Dict[str, System]:
    """Load systems from base data and create lookup dict"""
    systems = {}
    for s in base_data.get('systems', []):
        system = System(
            SystemID=s['SystemID'],
            DisplayName=s['DisplayName'],
            Class=s['Class'],
            BaseScale=s['BaseScale'],
            ScaleFactor=s['ScaleFactor'],
            MeasureName=s['MeasureName'],
            FractalDimension=s.get('FractalDimension'),
            TheoreticalLogLogSlope=s['TheoreticalLogLogSlope']
        )
        systems[system.SystemID] = system
    return systems


def load_test_scales(test_input: Dict) -> List[Scale]:
    """Load test scales from test input (raw facts only)"""
    scales = []
    for s in test_input.get('scales', []):
        scale = Scale(
            ScaleID=s['ScaleID'],
            System=s['System'],
            Iteration=s['Iteration'],
            Measure=s['Measure'],
            IsProjected=s.get('IsProjected', True)
        )
        scales.append(scale)
    return scales


def compute_derived_values(scale: Scale, systems: Dict[str, System]) -> Scale:
    """Compute all derived values for a scale"""
    system = systems.get(scale.System)
    if not system:
        raise ValueError(f"System not found: {scale.System}")
    
    # Lookup values from parent system
    scale.BaseScale = system.BaseScale
    scale.ScaleFactor = system.ScaleFactor
    
    # Calculated values
    scale.ScaleFactorPower = math.pow(scale.ScaleFactor, scale.Iteration)
    scale.Scale = scale.BaseScale * scale.ScaleFactorPower
    scale.LogScale = math.log10(scale.Scale) if scale.Scale > 0 else 0
    scale.LogMeasure = math.log10(scale.Measure) if scale.Measure > 0 else 0
    
    return scale


def round_for_comparison(value: float, decimals: int = 6) -> float:
    """Round a float for consistent comparison (6 decimal places)"""
    if value is None:
        return None
    return round(value, decimals)


def scale_to_dict(scale: Scale) -> Dict:
    """Convert Scale to dict with rounded values for JSON output (6 decimal places)"""
    return {
        'ScaleID': scale.ScaleID,
        'System': scale.System,
        'Iteration': scale.Iteration,
        'Measure': round_for_comparison(scale.Measure),
        'BaseScale': round_for_comparison(scale.BaseScale),
        'ScaleFactor': round_for_comparison(scale.ScaleFactor),
        'ScaleFactorPower': round_for_comparison(scale.ScaleFactorPower),
        'Scale': round_for_comparison(scale.Scale),
        'LogScale': round_for_comparison(scale.LogScale),
        'LogMeasure': round_for_comparison(scale.LogMeasure),
        'IsProjected': scale.IsProjected
    }


def system_to_dict(system: System) -> Dict:
    """Convert System dataclass to dict for visualization"""
    return {
        'SystemID': system.SystemID,
        'DisplayName': system.DisplayName,
        'Class': system.Class,
        'BaseScale': system.BaseScale,
        'ScaleFactor': system.ScaleFactor,
        'MeasureName': system.MeasureName,
        'FractalDimension': system.FractalDimension,
        'TheoreticalLogLogSlope': system.TheoreticalLogLogSlope
    }


def compare_values(expected, actual, tolerance: float = TOLERANCE) -> bool:
    """Compare two values with tolerance for floats"""
    if expected is None and actual is None:
        return True
    if expected is None or actual is None:
        return False
    if isinstance(expected, (int, float)) and isinstance(actual, (int, float)):
        return abs(expected - actual) < tolerance
    return expected == actual


def validate_results(computed_scales: List[Dict], answer_key: Dict) -> tuple:
    """Validate computed scales against answer key"""
    expected_by_id = {s['ScaleID']: s for s in answer_key.get('scales', [])}
    
    pass_count = 0
    fail_count = 0
    failures = []
    
    computed_fields = ['BaseScale', 'ScaleFactor', 'ScaleFactorPower', 'Scale', 'LogScale', 'LogMeasure']
    
    for scale in computed_scales:
        scale_id = scale.get('ScaleID')
        expected = expected_by_id.get(scale_id)
        if not expected:
            # Not a test scale - skip validation
            continue
        
        mismatches = []
        
        for field in computed_fields:
            exp_val = expected.get(field)
            act_val = scale.get(field)
            
            if not compare_values(exp_val, act_val):
                mismatches.append(f"{field}: expected {exp_val}, got {act_val}")
        
        if mismatches:
            fail_count += 1
            failures.append((scale_id, mismatches))
        else:
            pass_count += 1
    
    return pass_count, fail_count, failures


def main():
    """Main test runner"""
    # Check if test data exists
    base_data_path = TEST_DATA_DIR / 'base-data.json'
    test_input_path = TEST_DATA_DIR / 'test-input.json'
    answer_key_path = TEST_DATA_DIR / 'answer-key.json'
    
    if not base_data_path.exists() or not test_input_path.exists():
        print(f"{RED}Error: Test data not found. Run generate-test-data.py first.{RESET}")
        sys.exit(1)
    
    # Load data
    base_data = load_json(base_data_path)
    test_input = load_json(test_input_path)
    answer_key = load_json(answer_key_path)
    
    # Load systems
    systems = load_systems(base_data)
    
    # Load test scales (raw facts only)
    test_scales = load_test_scales(test_input)
    
    # Compute derived values for test scales
    computed_test_scales = []
    for scale in test_scales:
        computed = compute_derived_values(scale, systems)
        computed_test_scales.append(computed)
    
    # Convert to dicts for output
    computed_test_dicts = [scale_to_dict(s) for s in computed_test_scales]
    
    # Save results (test scales only for validation compatibility)
    results = {
        'platform': 'python',
        'timestamp': datetime.now(timezone.utc).isoformat(),
        'scales': computed_test_dicts
    }
    
    results_path = TEST_RESULTS_DIR / 'python-results.json'
    save_json(results_path, results)
    
    # Merge base scales with computed test scales for full visualization
    base_scales = base_data.get('scales', [])
    all_scales = merge_scales(base_scales, computed_test_dicts)
    
    # Validate against answer key
    pass_count, fail_count, failures = validate_results(computed_test_dicts, answer_key)
    
    # Convert systems to dict format for visualization
    systems_dict = {sid: system_to_dict(sys) for sid, sys in systems.items()}
    
    # Print full report using shared library
    print_full_report(
        platform='python',
        all_scales=all_scales,
        systems=systems_dict,
        pass_count=pass_count,
        fail_count=fail_count,
        failures=failures,
        show_plots=True
    )
    
    # Exit with appropriate code
    sys.exit(0 if fail_count == 0 else 1)


if __name__ == '__main__':
    main()
